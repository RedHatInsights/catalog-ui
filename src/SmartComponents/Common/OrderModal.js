import React, { Component } from 'react';
import { Button, Grid, Modal} from '@patternfly/react-core';
import { connect } from 'react-redux';
import './orderservice.scss';
import propTypes from 'prop-types';
import { OrderServiceFormSteps } from '../Order/OrderServiceFormConstants';
import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../../PresentationalComponents/Shared/ImageWithDefault';
import { Wizard } from 'patternfly-react';

// useless
const renderStepWizardPage = (componentPage, props) => {
  const StepComponent = componentPage;
  return (<StepComponent { ...props } />);
};

class OrderModal extends Component {
  state = {
    activeStepIndex: 0
  };

  onNext = () => {
    const { activeStepIndex } = this.state;
    const numberSteps = OrderServiceFormSteps.length;

    if (activeStepIndex < numberSteps - 1) {
      this.setState({ activeStepIndex: activeStepIndex + 1 });
    }
  }

  renderWizardSteps = () => {
    const { activeStepIndex } = this.state;
    const wizardSteps = OrderServiceFormSteps;
    const activeStep = wizardSteps[activeStepIndex];

    return wizardSteps.map((step, stepIndex) => (
      <Wizard.Step
        key={ stepIndex }
        stepIndex={ stepIndex }
        step={ step.step }
        label={ step.label }
        title={ step.title }
        activeStep={ activeStep && activeStep.step }
      />
    ));
  };

  render() {
    const showOrder = this.props.open;

    if (!showOrder) {
      return null;
    }

    const { activeStepIndex } = this.state;
    const wizardSteps = OrderServiceFormSteps;

    return (
      <div>
        <div>
          <ImageWithDefault src = { this.props.servicedata.imageUrl || CatItemSvg } width="40" />
          { this.props.servicedata.name }}
        </div>
        <Wizard.Steps steps={ this.renderWizardSteps(this.props.servicedata) }/>
        <Wizard.Row>
          <Wizard.Main>
            <Grid>
              { wizardSteps.map((step, stepIndex) => (
                <Wizard.Contents key={ step.title } stepIndex={ stepIndex } activeStepIndex={ activeStepIndex }>
                  { renderStepWizardPage(wizardSteps[stepIndex].page, this.props.servicedata) }
                </Wizard.Contents>
              )) }
              { activeStepIndex !== wizardSteps.length - 1 && (
                <div>
                  <br/>
                  <br/>
                  <Button variant="primary" type="button" onClick={ this.onNext }>
                    Order
                  </Button>
                </div>
              ) }
            </Grid>
          </Wizard.Main>
        </Wizard.Row>
      </div>
    );
  }
}

OrderModal.propTypes = {
  orderData: propTypes.func,
  showOrder: propTypes.bool,
  servicedata: propTypes.object,
  stepParametersValid: propTypes.bool,
  fulfilled: propTypes.bool,
  error: propTypes.bool,
  open: propTypes.bool
};

const mapStateToProps = ({ orderReducer: { isLoading, selectedItem, servicePlans }}) => ({
  isLoading,
  selectedItem,
  servicePlans
});

export default connect(mapStateToProps)(OrderModal);
