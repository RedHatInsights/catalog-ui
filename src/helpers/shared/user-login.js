import axios from 'axios';
import { RequestApi } from '@redhat-cloud-services/approval-client';
import { DefaultApi as SourcesDefaultApi } from '@redhat-cloud-services/sources-client';
import { DefaultApi as TopologicalDefaultApi } from '@redhat-cloud-services/topological-inventory-client';
import { PortfolioApi, PortfolioItemApi, OrderApi, OrderItemApi, IconApi } from '@redhat-cloud-services/catalog-client';

import { SOURCES_API_BASE, TOPOLOGICAL_INVENTORY_API_BASE, CATALOG_API_BASE, APPROVAL_API_BASE, RBAC_API_BASE } from '../../utilities/constants';
import { AccessApi, PrincipalApi, GroupApi } from '@redhat-cloud-services/rbac-client';
import { stringify } from 'qs';

const axiosInstance = axios.create({
  paramsSerializer: params => stringify(params)
});

const resolveInterceptor = response => response.data || response;
const errorInterceptor = (error = {}) => {
  throw { ...error.response };
};

// check identity before each request. If the token is expired it will log out user
axiosInstance.interceptors.request.use(async config => {
  await window.insights.chrome.auth.getUser();
  return config;
});
axiosInstance.interceptors.response.use(resolveInterceptor);
axiosInstance.interceptors.response.use(null, errorInterceptor);

const orderApi = new OrderApi(undefined, CATALOG_API_BASE, axiosInstance);
const orderItemApi = new OrderItemApi(undefined, CATALOG_API_BASE, axiosInstance);
const portfolioApi = new PortfolioApi(undefined, CATALOG_API_BASE, axiosInstance);
const portfolioItemApi = new PortfolioItemApi(undefined, CATALOG_API_BASE, axiosInstance);
const requestsApi = new RequestApi(undefined, APPROVAL_API_BASE, axiosInstance);
const sourcesApi = new SourcesDefaultApi(undefined, SOURCES_API_BASE, axiosInstance);
const topologicalInventoryApi = new TopologicalDefaultApi(undefined, TOPOLOGICAL_INVENTORY_API_BASE, axiosInstance);
const iconApi = new IconApi(undefined, CATALOG_API_BASE, axiosInstance);

export function getSourcesApi() {
  return sourcesApi;
}

export function getTopologocalInventoryApi() {
  return topologicalInventoryApi;
}

export function getPortfolioApi() {
  return portfolioApi;
}

export function getPortfolioItemApi() {
  return portfolioItemApi;
}

export function getOrderApi() {
  return orderApi;
}

export function getOrderItemApi() {
  return orderItemApi;
}

export function getRequestsApi() {
  return requestsApi;
}

let rbacAccessApi = new AccessApi(undefined, RBAC_API_BASE, axiosInstance);
let rbacPrincipalApi = new PrincipalApi(undefined, RBAC_API_BASE, axiosInstance);
let rbacGroupApi = new GroupApi(undefined, RBAC_API_BASE, axiosInstance);

export function getRbacAccessApi() {
  return rbacAccessApi;
}

export function getRbacPrincipalApi() {
  return rbacPrincipalApi;
}

export function getRbacGroupApi() {
  return rbacGroupApi;
}

export function getAxiosInstance() {
  return axiosInstance;
}

export function getIconApi() {
  return iconApi;
}

const grapqlInstance = axios.create();
grapqlInstance.interceptors.request.use(async config => {
  await window.insights.chrome.auth.getUser();
  return config;
});
/**
 * Graphql does not return error response when the qery fails.
 * Instead it returns 200 response with error object.
 * We catch it and throw it to trigger notification middleware
 */
grapqlInstance.interceptors.response.use(({ data }) => {
  if (data.errors) {
    throw {
      message: data.errors[0].errorType,
      data: data.errors[0].message
    };
  }

  return data;
});

export function getGraphqlInstance() {
  return grapqlInstance;
}
