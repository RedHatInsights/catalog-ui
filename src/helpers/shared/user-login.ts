import axios, { AxiosResponse, AxiosRequestConfig, AxiosInstance } from 'axios';
import {
  RequestApi,
  WorkflowApi
} from '@redhat-cloud-services/approval-client';
import {
  PortfolioApi,
  PortfolioItemApi,
  OrderApi,
  OrderItemApi,
  IconApi,
  ServicePlansApi,
  OrderProcessApi
} from '@redhat-cloud-services/catalog-client';

import {
  CATALOG_API_BASE,
  APPROVAL_API_BASE,
  RBAC_API_BASE
} from '../../utilities/constants';
import { GroupApi } from '@redhat-cloud-services/rbac-client';
import { stringify } from 'qs';

export interface ApiHeaders extends Headers {
  'x-rh-insights-request-id': string;
}

export interface ErrorResponse {
  headers?: ApiHeaders;
}

export interface ServerError {
  response?: ErrorResponse;
  status?: 403 | 404 | 401 | 400 | 429 | 500 | 200; // not a complete list, replace by library with complete interface
  config?: AxiosRequestConfig;
}

const createAxiosInstance = () => {
  if (window.catalog?.standalone) {
    const token = window.catalog?.token;
    return axios.create({
      paramsSerializer: (params) => stringify(params),
      headers: { Authorization: `Basic ${token}` }
    });
  } else {
    return axios.create({
      paramsSerializer: (params) => stringify(params)
    });
  }
};

const axiosInstance: AxiosInstance = createAxiosInstance();

const resolveInterceptor = (response: AxiosResponse) =>
  response.data || response;
const errorInterceptor = (error: ServerError = {}) => {
  const requestId = error.response?.headers?.['x-rh-insights-request-id'];
  throw requestId ? { ...error.response, requestId } : { ...error.response };
};

const unauthorizedInterceptor = (error: ServerError = {}) => {
  if (error.status === 403) {
    throw {
      ...error,
      redirect: {
        pathname: '/403',
        message: error.config?.url
      }
    };
  }

  throw error;
};

// check identity before each request. If the token is expired it will log out user
axiosInstance.interceptors.request.use(async (config) => {
  // eslint-disable-next-line no-undef
  if (!window.catalog?.standalone) {
    await window.insights.chrome.auth.getUser();
  } else {
    if (window.catalog?.token) {
      config.headers = { Authorization: `Basic ${window.catalog.token}` };
      //config.headers.Authorization = `Basic ${window.catalog.token}`;
    }
  }

  return config;
});
axiosInstance.interceptors.response.use(resolveInterceptor);
axiosInstance.interceptors.response.use(undefined, errorInterceptor);
axiosInstance.interceptors.response.use(undefined, unauthorizedInterceptor);

const orderApi = new OrderApi(undefined, CATALOG_API_BASE, axiosInstance);
const orderItemApi = new OrderItemApi(
  undefined,
  CATALOG_API_BASE,
  axiosInstance
);
const portfolioApi = new PortfolioApi(
  undefined,
  CATALOG_API_BASE,
  axiosInstance
);
const portfolioItemApi = new PortfolioItemApi(
  undefined,
  CATALOG_API_BASE,
  axiosInstance
);
const requestsApi = new RequestApi(undefined, APPROVAL_API_BASE, axiosInstance);
const workflowApi = new WorkflowApi(
  undefined,
  APPROVAL_API_BASE,
  axiosInstance
);
const iconApi = new IconApi(undefined, CATALOG_API_BASE, axiosInstance);
const servicePlansApi = new ServicePlansApi(
  undefined,
  CATALOG_API_BASE,
  axiosInstance
);
const orderProcessApi = new OrderProcessApi(
  undefined,
  CATALOG_API_BASE,
  axiosInstance
);

export function getPortfolioApi(): PortfolioApi {
  return portfolioApi;
}

export function getPortfolioItemApi(): PortfolioItemApi {
  return portfolioItemApi;
}

export function getOrderApi(): OrderApi {
  return orderApi;
}

export function getOrderItemApi(): OrderItemApi {
  return orderItemApi;
}

export function getRequestsApi(): RequestApi {
  return requestsApi;
}

const rbacGroupApi = new GroupApi(undefined, RBAC_API_BASE, axiosInstance);

export function getRbacGroupApi(): GroupApi {
  return rbacGroupApi;
}

export function getWorkflowApi(): WorkflowApi {
  return workflowApi;
}

export function getAxiosInstance(): AxiosInstance {
  return axiosInstance;
}

export function getIconApi(): IconApi {
  return iconApi;
}

export function getServicePlansApi(): ServicePlansApi {
  return servicePlansApi;
}

export function getOrderProcessApi(): OrderProcessApi {
  return orderProcessApi;
}

const grapqlInstance = axios.create();

grapqlInstance.interceptors.request.use(async (config) => {
  if (!window.catalog?.standalone) {
    await window.insights.chrome.auth.getUser();
  }

  return config;
});
/**
 * Graphql does not return error response when the query fails.
 * Instead it returns 200 response with error object.
 * We catch it and throw it to trigger the notification middleware
 */
grapqlInstance.interceptors.response.use(({ data }) => {
  if (data.errors) {
    throw {
      message: data.errors[0].errorType,
      data: data.errors[0].message
    };
  }

  return data;
});

export function getGraphqlInstance(): AxiosInstance {
  return grapqlInstance;
}
