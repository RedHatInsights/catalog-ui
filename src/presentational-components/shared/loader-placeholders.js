import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '@patternfly/react-core/dist/js/components/Spinner/Spinner';
import { Section } from '@redhat-cloud-services/frontend-components/components/cjs/Section';
import {
  Bullseye,
  Card,
  CardBody,
  DataList,
  DataListCell,
  DataListItem,
  DataListItemRow,
  DataListItemCells,
  Grid,
  GridItem,
  Gallery,
  GalleryItem,
  Form,
  FormGroup,
  TextContent,
  Text,
  TextVariants,
  ActionGroup,
  Button
} from '@patternfly/react-core';
import styled, { keyframes } from 'styled-components';
import { StyledToolbar } from '../styled-components/toolbars';

const wave = keyframes`
  0% {
    transform: translateX(-100%);
  }
  60% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(100%);
  }
`;

const Skeleton = ({ component: Component = 'span', ...props }) => {
  return (
    <SkeletonContainer {...props}>
      <Component />
    </SkeletonContainer>
  );
};

Skeleton.propTypes = {
  component: PropTypes.string
};

const SkeletonContainer = styled.div`
  & > * {
    position: relative;
    overflow: hidden;
    width: ${({ width }) =>
      width ? `${width}${typeof width === 'number' ? 'px' : ''}` : '100%'};
    height: ${({ height }) =>
      height ? `${height}${typeof height === 'number' ? 'px' : ''}` : '20px'};
    display: block;
    border-radius: 3px;
    background-color: ${({ secondaryColor }) =>
      secondaryColor ? secondaryColor : '#f3f3f3'};
    &:after {
      animation: 2s ${wave} linear 0.5s infinite;
      background: linear-gradient(90deg, transparent, #ecebeb, transparent);
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      right: 0;
      top: 0;
      transform: translateX(-100%);
      z-index: 1;
    }
  }
`;

export const CardLoader = ({ items }) => (
  <Grid>
    <GridItem sm={12}>
      <Section type="content">
        <Gallery hasGutter>
          {[...Array(items)].map((_item, index) => (
            <GalleryItem key={index}>
              <Card style={{ height: 350 }}>
                <CardBody>
                  <Skeleton height={70} width="85%" className="pf-u-mb-lg" />
                  <Skeleton height={5} width="90%" className="pf-u-mb-sm" />
                  <Skeleton height={5} width="100%" className="pf-u-mb-sm" />
                  <Skeleton height={5} width="76%" className="pf-u-mb-sm" />
                </CardBody>
              </Card>
            </GalleryItem>
          ))}
        </Gallery>
      </Section>
    </GridItem>
  </Grid>
);

CardLoader.propTypes = {
  items: PropTypes.number
};

CardLoader.defaultProps = {
  items: 13
};

export const AppPlaceholder = () => (
  <section className="pf-u-m-0 pf-u-p-0 pf-l-page__main-section pf-c-page__main-section">
    <Skeleton height={32} className="pf-u-p-lg global-primary-background" />
    <div className="pf-u-mt-lg">
      <Bullseye>
        <Spinner />
      </Bullseye>
    </div>
  </section>
);

export const ToolbarTitlePlaceholder = () => <Skeleton height={30} />;

const ProducLoaderColumn = styled.div`
  width: 100%;
  max-width: 250px;
`;

export const ProductLoaderPlaceholder = () => (
  <Fragment>
    <Skeleton height={70} className="pf-u-mb-xl" />
    <ProducLoaderColumn>
      {[...Array(3)].map((_, index) => (
        <Fragment key={index}>
          <Skeleton height={8} className="pf-u-mb-sm" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-sm" width="60%" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-sm" width="50%" />
          <Skeleton height={8} className="pf-u-mb-sm" width="80%" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-lg" width="40%" />
        </Fragment>
      ))}
    </ProducLoaderColumn>
  </Fragment>
);

export const IconPlaceholder = ({ height }) => (
  <svg height={height} width={height}>
    <circle cx={height / 2} cy={height / 2} r={height / 2} fill="#ecebeb" />
  </svg>
);

IconPlaceholder.propTypes = {
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

IconPlaceholder.defaultProps = {
  height: '40'
};

const FormItemLoader = () => <Skeleton height={38} className="pf-u-mb-lg" />;

export const ShareLoader = () => (
  <Form>
    <TextContent>
      <Text component={TextVariants.small}>Invite group</Text>
    </TextContent>
    <FormGroup fieldId="1">
      <FormItemLoader />
    </FormGroup>
    <TextContent>
      <Text component={TextVariants.small}>Groups with access</Text>
    </TextContent>
    <FormGroup fieldId="3">
      <FormItemLoader />
    </FormGroup>
    <FormGroup fieldId="4">
      <FormItemLoader />
    </FormGroup>
  </Form>
);

export const WorkflowLoader = () => (
  <Form>
    <FormGroup fieldId="2" label="Select workflow">
      <FormItemLoader />
    </FormGroup>
    <ActionGroup>
      <Button variant="primary" isDisabled>
        Save
      </Button>
    </ActionGroup>
  </Form>
);

export const ListLoader = ({ items }) => (
  <DataList aria-label="list-loader" aria-labelledby="datalist-placeholder">
    {[...Array(items)].map((_item, index) => (
      <DataListItem key={index} aria-labelledby="datalist-item-placeholder">
        <DataListItemRow aria-label="datalist-item-placeholder-row">
          <DataListItemCells
            dataListCells={[
              <DataListCell key="1">
                <Skeleton height={67} />
              </DataListCell>
            ]}
          />
        </DataListItemRow>
      </DataListItem>
    ))}
  </DataList>
);

ListLoader.propTypes = {
  items: PropTypes.number
};

ListLoader.defaultProps = {
  items: 5
};

export const OrderDetailToolbarPlaceholder = () => <Skeleton height={70} />;

export const PlatformToolbarPlaceholder = () => (
  <StyledToolbar className="pf-u-pr-lg pf-u-pl-lg pf-u-pb-md pf-u-pt-md">
    <Skeleton height={36} width={300} />
  </StyledToolbar>
);
