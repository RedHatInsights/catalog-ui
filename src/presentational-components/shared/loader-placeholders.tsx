/* eslint-disable react/prop-types */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '@patternfly/react-core/dist/js/components/Spinner/Spinner';
import { Section } from '@redhat-cloud-services/frontend-components/Section';
import {
  Bullseye,
  Card,
  CardBody,
  DataList,
  DataListCell,
  DataListItem,
  DataListItemRow,
  DataListItemCells,
  Grid,
  GridItem,
  Gallery,
  GalleryItem,
  Form,
  FormGroup,
  TextContent,
  Text,
  TextVariants,
  ActionGroup,
  Button
} from '@patternfly/react-core';
import styled, { keyframes } from 'styled-components';
import { StyledToolbar } from '../styled-components/toolbars';
import actionMessages from '../../messages/actions.messages';
import formsMessages from '../../messages/forms.messages';
import useFormatMessage from '../../utilities/use-format-message';

const wave = keyframes`
  0% {
    transform: translateX(-100%);
  }
  60% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(100%);
  }
`;

const Skeleton = ({ component: Component = 'span', ...props }) => {
  return (
    <SkeletonContainer {...props}>
      <Component />
    </SkeletonContainer>
  );
};

Skeleton.propTypes = {
  component: PropTypes.string
};

const SkeletonContainer = styled.div<{
  width?: number | string;
  height?: number | string;
  secondaryColor?: string;
}>`
  & > * {
    position: relative;
    overflow: hidden;
    width: ${({ width }) =>
      width ? `${width}${typeof width === 'number' ? 'px' : ''}` : '100%'};
    height: ${({ height }) =>
      height ? `${height}${typeof height === 'number' ? 'px' : ''}` : '20px'};
    display: block;
    border-radius: 3px;
    background-color: ${({ secondaryColor }) =>
      secondaryColor ? secondaryColor : '#f3f3f3'};
    &:after {
      animation: 2s ${wave} linear 0.5s infinite;
      background: linear-gradient(90deg, transparent, #ecebeb, transparent);
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      right: 0;
      top: 0;
      transform: translateX(-100%);
      z-index: 1;
    }
  }
`;

export interface CardLoaderProps {
  items?: number;
}
export const CardLoader: React.ComponentType<CardLoaderProps> = ({
  items = 13
}) => (
  <Grid hasGutter>
    <GridItem sm={12}>
      <Section type="content">
        <Gallery hasGutter>
          {[...Array(items)].map((_item, index) => (
            <GalleryItem key={index}>
              <Card style={{ height: 350 }}>
                <CardBody>
                  <Skeleton height={70} width="85%" className="pf-u-mb-lg" />
                  <Skeleton height={5} width="90%" className="pf-u-mb-sm" />
                  <Skeleton height={5} width="100%" className="pf-u-mb-sm" />
                  <Skeleton height={5} width="76%" className="pf-u-mb-sm" />
                </CardBody>
              </Card>
            </GalleryItem>
          ))}
        </Gallery>
      </Section>
    </GridItem>
  </Grid>
);

export const AppPlaceholder: React.ElementType = () => (
  <section className="pf-u-m-0 pf-u-p-0 pf-l-page__main-section pf-c-page__main-section">
    <Skeleton height={32} className="pf-u-p-lg global-primary-background" />
    <div className="pf-u-mt-lg">
      <Bullseye>
        <Spinner />
      </Bullseye>
    </div>
  </section>
);

export const ToolbarTitlePlaceholder: React.ElementType = () => (
  <Skeleton height={30} />
);

const ProducLoaderColumn = styled.div`
  width: 100%;
  max-width: 250px;
`;

export const ProductLoaderPlaceholder: React.ElementType = () => (
  <Fragment>
    <Skeleton height={70} className="pf-u-mb-xl" />
    <ProducLoaderColumn>
      {[...Array(3)].map((_, index) => (
        <Fragment key={index}>
          <Skeleton height={8} className="pf-u-mb-sm" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-sm" width="60%" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-sm" width="50%" />
          <Skeleton height={8} className="pf-u-mb-sm" width="80%" />
          <Skeleton height={8} className="pf-u-ml-md pf-u-mb-lg" width="40%" />
        </Fragment>
      ))}
    </ProducLoaderColumn>
  </Fragment>
);

export interface IconPlaceholderProps {
  height: number;
}
export const IconPlaceholder: React.ElementType<IconPlaceholderProps> = ({
  height
}) => (
  <svg height={height} width={height}>
    <circle cx={height / 2} cy={height / 2} r={height / 2} fill="#ecebeb" />
  </svg>
);

export const FormItemLoader: React.ElementType = () => (
  <Skeleton height={38} className="pf-u-mb-lg" />
);

export const ShareLoader: React.ElementType = () => {
  const formatMessage = useFormatMessage();
  return (
    <Form>
      <TextContent>
        <Text component={TextVariants.small}>
          {formatMessage(formsMessages.inviteGroup)}
        </Text>
      </TextContent>
      <FormGroup fieldId="1">
        <FormItemLoader />
      </FormGroup>
      <TextContent>
        <Text component={TextVariants.small}>
          {formatMessage(formsMessages.groupsAccess)}
        </Text>
      </TextContent>
      <FormGroup fieldId="3">
        <FormItemLoader />
      </FormGroup>
      <FormGroup fieldId="4">
        <FormItemLoader />
      </FormGroup>
    </Form>
  );
};

export const WorkflowLoader: React.ElementType = () => {
  const formatMessage = useFormatMessage();
  return (
    <Form>
      <FormGroup fieldId="2">
        <FormItemLoader />
      </FormGroup>
      <ActionGroup>
        <Button variant="primary" isDisabled>
          {formatMessage(actionMessages.save)}
        </Button>
      </ActionGroup>
    </Form>
  );
};

export interface ListLoaderProps {
  items?: number;
}
export const ListLoader: React.ElementType<ListLoaderProps> = ({
  items = 5
}) => (
  <DataList aria-label="list-loader" aria-labelledby="datalist-placeholder">
    {[...Array(items)].map((_item, index) => (
      <DataListItem key={index} aria-labelledby="datalist-item-placeholder">
        <DataListItemRow aria-label="datalist-item-placeholder-row">
          <DataListItemCells
            dataListCells={[
              <DataListCell key="1">
                <Skeleton height={67} />
              </DataListCell>
            ]}
          />
        </DataListItemRow>
      </DataListItem>
    ))}
  </DataList>
);

export const OrderDetailToolbarPlaceholder: React.ElementType = () => (
  <Skeleton height={70} />
);

export const PlatformToolbarPlaceholder: React.ElementType = () => (
  <StyledToolbar className="pf-u-pr-lg pf-u-pl-lg pf-u-pb-md pf-u-pt-md">
    <Skeleton height={36} width={300} />
  </StyledToolbar>
);
