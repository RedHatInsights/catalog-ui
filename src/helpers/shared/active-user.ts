/* eslint camelcase: 0 */
import { getAxiosInstance } from '../shared/user-login';
import { AUTH_API_BASE } from '../../utilities/constants';
const axiosInstance = getAxiosInstance();

export const getUser = () => axiosInstance.get(`${AUTH_API_BASE}/me/`);

export const logoutUser = () => {
  localStorage.removeItem('catalog_user');
  return axiosInstance.post(`${AUTH_API_BASE}/logout/`);
};

export const loginUser = () => {
  window.location.replace(`${AUTH_API_BASE}/login/`);
  return null;
};
