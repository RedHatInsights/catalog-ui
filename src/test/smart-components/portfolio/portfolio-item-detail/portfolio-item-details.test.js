import React from 'react';
import thunk from 'redux-thunk';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import OrderModal from '../../../../smart-components/common/order-modal';
import { ProductLoaderPlaceholder } from '../../../../presentational-components/shared/loader-placeholders';
import ItemDetailInfoBar from '../../../../smart-components/portfolio/portfolio-item-detail/item-detail-info-bar';
import PortfolioItemDetail from '../../../../smart-components/portfolio/portfolio-item-detail/portfolio-item-detail';
import {
  CATALOG_API_BASE,
  SOURCES_API_BASE
} from '../../../../utilities/constants';
import ItemDetailDescription from '../../../../smart-components/portfolio/portfolio-item-detail/item-detail-description';
import { PortfolioItemDetailToolbar } from '../../../../smart-components/portfolio/portfolio-item-detail/portfolio-item-detail-toolbar';
import { mockApi } from '../../../__mocks__/user-login';
import { Alert } from '@patternfly/react-core';

describe('<PortfolioItemDetail />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const ComponentWrapper = ({
    store,
    children,
    initialEntries,
    initialIndex
  }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
        {children}
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {};
    initialState = {
      breadcrumbsReducer: { fragments: [] },
      platformReducer: {
        platforms: [],
        platformIconMapping: {}
      },
      portfolioReducer: {
        portfolioItem: {
          portfolioItem: {
            id: '123',
            service_offering_source_ref: '111',
            created_at: '123',
            name: 'bar',
            metadata: {
              user_capabilities: {
                copy: true,
                update: true,
                destroy: true
              }
            }
          },
          source: { id: '111', name: 'Source name' },
          portfolio: {
            name: 'Portfolio name'
          }
        }
      },
      orderReducer: {
        servicePlans: [{ create_json_schema: { schema: { fields: [] } } }]
      }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const wrapper = shallow(
      <ComponentWrapper store={mockStore({})}>
        <PortfolioItemDetail {...initialProps} />
      </ComponentWrapper>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should mount correct component before and after load', async (done) => {
    const store = mockStore(initialState);
    mockApi.onGet(`${CATALOG_API_BASE}/portfolios/123`).replyOnce(200, {});
    mockApi.onGet(`${CATALOG_API_BASE}/portfolio_items/321`).replyOnce(200, {});
    mockApi.onGet(`${SOURCES_API_BASE}/sources/source-id`).replyOnce(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper
          initialEntries={[
            '/portfolio/portfolio-item?source=source-id&portfolio=123&portfolio-item=321'
          ]}
          store={store}
        >
          <PortfolioItemDetail {...initialProps} />
        </ComponentWrapper>
      );
      expect(wrapper.find(ProductLoaderPlaceholder)).toHaveLength(1);
    });
    wrapper.update();
    expect(wrapper.find(ItemDetailInfoBar)).toHaveLength(1);
    expect(wrapper.find(ItemDetailDescription)).toHaveLength(1);
    expect(wrapper.find(PortfolioItemDetailToolbar)).toHaveLength(1);
    done();
  });

  it('should mount and open order modal', async (done) => {
    const store = mockStore(initialState);

    mockApi.onGet(`${CATALOG_API_BASE}/portfolios/123`).replyOnce(200, {});
    mockApi.onGet(`${CATALOG_API_BASE}/portfolio_items/321`).replyOnce(200, {});
    mockApi
      .onGet(`${CATALOG_API_BASE}/portfolio_items/123/service_plans`)
      .replyOnce(200, {});
    mockApi
      .onGet(
        `${CATALOG_API_BASE}/portfolio_items/321/provider_control_parameters`
      )
      .replyOnce(200, {
        properties: { namespace: { enum: [] } }
      }),
      mockApi.onGet(`${SOURCES_API_BASE}/sources/source-id`).replyOnce(200, {});

    let wrapper;

    await act(async () => {
      wrapper = mount(
        <ComponentWrapper
          store={store}
          initialEntries={[
            '/portfolio/portfolio-item?source=source-id&portfolio=123&portfolio-item=321',
            '/portfolio/portfolio-item/order?source=source-id&portfolio=123&portfolio-item=321'
          ]}
          initialIndex={0}
        >
          <Route path="/portfolio/portfolio-item">
            <PortfolioItemDetail {...initialProps} />
          </Route>
        </ComponentWrapper>
      );
    });
    wrapper.update();
    wrapper
      .find(MemoryRouter)
      .instance()
      .history.push(
        '/portfolio/portfolio-item/order?source=source-id&portfolio=123&portfolio-item=321'
      );
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper.find(OrderModal)).toHaveLength(1);
    done();
  });

  it('should mount with missing source', async (done) => {
    const store = mockStore({
      ...initialState,
      portfolioReducer: {
        ...initialState.portfolioReducer,
        portfolioItem: {
          ...initialState.portfolioReducer.portfolioItem,
          source: {
            notFound: true
          }
        }
      }
    });
    mockApi.onGet(`${CATALOG_API_BASE}/portfolios/123`).replyOnce(200, {});
    mockApi.onGet(`${CATALOG_API_BASE}/portfolio_items/321`).replyOnce(200, {});
    mockApi.onGet(`${SOURCES_API_BASE}/sources/source-id`).replyOnce(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper
          initialEntries={[
            '/portfolio/portfolio-item?source=source-id&portfolio=123&portfolio-item=321'
          ]}
          store={store}
        >
          <PortfolioItemDetail {...initialProps} />
        </ComponentWrapper>
      );
    });
    wrapper.update();
    expect(wrapper.find(Alert)).toHaveLength(1);
    done();
  });
});
