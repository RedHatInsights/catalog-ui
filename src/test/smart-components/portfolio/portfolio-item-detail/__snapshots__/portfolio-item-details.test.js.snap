// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PortfolioItemDetail /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "permissions": Array [],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <MemoryRouter>
      <PortfolioItemDetail />
    </MemoryRouter>
  </Provider>
  .
</ContextProvider>
`;
